### PR TITLE
Add e2e HPA Behavior Tests: Scale up and down controls

### LINKED ISSUE #102369: Create HPA Behavior E2E Test
#### What would you like to be added:

We need an E2E test for HPA behavior.

#### Why is this needed:

HPA v2beta2 is graduating to stable!  

- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-autoscaling/2702-graduate-hpa-api-to-GA
- Tracking issue: https://github.com/kubernetes/enhancements/issues/2702
- Working doc: https://docs.google.com/document/d/1QGT9Waxc500ICGX_TGIoxfOWKmWbXYC6TWxfdUC2Njg/edit#
- Slack channel: #sig-autoscaling-api

#### What to do:

1. Read [End-to-End Testing in Kubernetes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md).
2. Tell @josephburnett on #sig-autoscaling-api "Hey, I want to do this." (just so two people don't start the same work)
3. Write a new E2E test that tests `spec.behavior`.

#### Details:

There are currently two kinds of HPA tests in the Kubernetes E2E suite.  One tests [CPU autoscaling](https://github.com/kubernetes/kubernetes/blob/c495744436fc94ebbef2fcbeb97699ca96fe02dd/test/e2e/autoscaling/horizontal_pod_autoscaling.go#L31) and the other tests [custom metrics autoscaling](https://github.com/kubernetes/kubernetes/blob/c495744436fc94ebbef2fcbeb97699ca96fe02dd/test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go#L48).

The [Behavior](https://pkg.go.dev/k8s.io/api/autoscaling/v2beta2#HorizontalPodAutoscalerSpec) feature is independent of the kind of metrics HPA is using.  After HPA calculates scale for all metrics, it select the maximum and then applies user configured limits which are `MinReplicas`, `MaxReplicas` and `Behavior`.  Behavior specifies a scale up and down limit in rate (relative and absolute) as well as a delay.

Defaults are:
```yaml
# Scale up 4 pods or 2x, whichever is larger.
# Delay scale down by 5 minutes.
spec:
  behavior:
    scaleUp:
      selectPolicy: "Max"
      policies:
      - type: "Pods"
        value: 4
        periodSeconds: 15
      - type: "Percent"
        value: 200
        periodSeconds: 15
    scaleDown:
      stabilizationWindowSeconds: 300
```

New behavior tests could be part of the CPU or the custom metrics suites, it doesn't really matter.  Whichever is easier to implement and has the least flakes.

Cases we should test:
1. Faster scale down works.  Set `behavior.scaleDown.stabilizationWindowSeconds: 30`, verify scale up, then verify scale down happens within 1 minute.
2. Scale up delay works.  Set `behavior.scaleUp.stabilizationWindowSeconds: 300`, continuously send high usage and verify scale up happens only after 5 minutes.
3. Disabling autoscaling works.  Set `behavior.scaleUp.selectPolicy: "Disabled"` and `behavior.scaleDown.selectPolicy: "Disabled"`.  This will the replacement for the special 0 replicas case when we deprecate it, we should make sure it reliably works.

Send a pull request (one squashed commit please) and notify #sig-autoscaling-api so we can take a look too. It will probably need API review.

Feel free to reach out for help on #sig-autoscaling-api if you have any questions!

### PR DESCRIPTION
**Pull request by piotrnosek**
#### What type of PR is this?
e2e test

#### What this PR does / why we need it:

Implementing e2e tests for HPA behavior, requested by https://github.com/kubernetes/kubernetes/issues/102369.

Specifically, this PR implements scale up/scale down controls: see also https://github.com/kubernetes/kubernetes/pull/97348.

It also refactors some common autoscaling utils.

#### Which issue(s) this PR fixes:

Fixes #102369

#### Does this PR introduce a user-facing change?

```release-note
    NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go [hunk 1]
```diff
@@ -378,4 +379,74 @@ var _ = SIGDescribe("[Feature:HPA] [Serial] [Slow] Horizontal pod autoscaling (n
 			framework.ExpectEqual(timeWaitedFor2 < deadline, true, "waited %s, wanted less than %s", timeWaitedFor2, deadline)
 		})
 	})
+
+	ginkgo.Describe("with both scale up and down controls configured", func() {
+		ginkgo.It("should keep recommendation within the range over two stabilization windows", func() {
```
Comment (pbetkier) [inline test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go:384]: I don't quite understand what does this scenario verify. We already have a test for scale up stabilization window, right? What's the difference here?

Comment (piotrnosek) [reply test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go:384]: Have a look at this issue for context: https://github.com/kubernetes/kubernetes/issues/96671.

This first test is reproducing error scenario 1 from that issue:
```
stabilizationWindowSeconds set up to 5 minutes for both scale up and scale down
min replica count 2
Application consumes big amount of resources during warm-up
0 minute: current replica count: 2
1 minute: desired replica count: 3 - no scale up stabilization window worked
2 minute: desired replica count: 1 (!) - scale up to 3 (max number from previous recommendations during stabilization window period)
```

This was a HPA problem when both up and down scale stabilisation is configured. There is a nice explanation with images from @josephburnett here: https://github.com/kubernetes/kubernetes/issues/96671#issuecomment-746283934

Comment (piotrnosek) [reply test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go:384]: And the requirement for that test was mentioned by Joe here: https://github.com/kubernetes/kubernetes/issues/110172 under this point:
```scale and and down controls (see https://github.com/kubernetes/kubernetes/pull/97348))```

Comment (pbetkier) [reply test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go:384]: Oh, thanks for explaining.

But you're not really verifying that the recommendation is kept within range, right? Just how long it took to scale to 6 pods.

**Review by pbetkier — APPROVED** [review]

File: test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go (hunk 1 above)
Comment (piotrnosek) [reply test/e2e/autoscaling/horizontal_pod_autoscaling_behavior.go:384]: After consideration - I've modifed this to be more precise:
1. We increase CPU usage and check if replicas stay in the desired window.
2. We wait for the scale up to finish after the window is over.
3. We decrease CPU usage and check if replicas stay in desired window.
4. We wait for the scale down to finish after window is over.

With such approach we verify both that the range is kept and that target replicas are met.

**Review by mwielgus — APPROVED** [review]
/lgtm
/approve

**Pull request merged**